File: cpp/src/gbm.cpp
// gbm.cpp -- implementation of gbm.hpp.

#include "gbm.hpp"

#include <cmath>
#include <stdexcept>

namespace quant {

// =====================================================================
// Anonymous-namespace helpers (file-local)
// =====================================================================

namespace {

// Draw a standard normal from rng by inversion.
//
// The distribution is constructed as a local object on each call to
// preserve byte-exact reproducibility with the inline implementation
// of Block 1.1's simulate_terminal_gbm: a per-call local distribution
// is guaranteed to be in its initial state, whereas a thread_local
// one could in principle carry residual state between calls.
double standard_normal(std::mt19937_64& rng) {
    std::uniform_real_distribution<double> uniform(0.0, 1.0);
    const double u = uniform(rng);
    return inverse_normal_cdf(u);
}

}  // anonymous namespace


// =====================================================================
// Public: inverse_normal_cdf (Acklam's approximation)
// =====================================================================

double inverse_normal_cdf(double p) {
    if (!(p > 0.0 && p < 1.0)) {
        throw std::invalid_argument(
            "inverse_normal_cdf: p must be in (0, 1)");
    }

    // Coefficients of the rational approximation. Relative error
    // below 1.15e-9 over (0, 1). Reference: P. J. Acklam (2003).
    constexpr double a[6] = {
        -3.969683028665376e+01,  2.209460984245205e+02,
        -2.759285104469687e+02,  1.383577518672690e+02,
        -3.066479806614716e+01,  2.506628277459239e+00
    };
    constexpr double b[5] = {
        -5.447609879822406e+01,  1.615858368580409e+02,
        -1.556989798598866e+02,  6.680131188771972e+01,
        -1.328068155288572e+01
    };
    constexpr double c[6] = {
        -7.784894002430293e-03, -3.223964580411365e-01,
        -2.400758277161838e+00, -2.549732539343734e+00,
         4.374664141464968e+00,  2.938163982698783e+00
    };
    constexpr double d[4] = {
         7.784695709041462e-03,  3.224671290700398e-01,
         2.445134137142996e+00,  3.754408661907416e+00
    };

    constexpr double p_low  = 0.02425;
    constexpr double p_high = 1.0 - p_low;

    if (p < p_low) {
        const double q = std::sqrt(-2.0 * std::log(p));
        return (((((c[0]*q + c[1])*q + c[2])*q + c[3])*q + c[4])*q + c[5])
             / ((((d[0]*q + d[1])*q + d[2])*q + d[3])*q + 1.0);
    }
    if (p <= p_high) {
        const double q = p - 0.5;
        const double r = q * q;
        return (((((a[0]*r + a[1])*r + a[2])*r + a[3])*r + a[4])*r + a[5]) * q
             / (((((b[0]*r + b[1])*r + b[2])*r + b[3])*r + b[4])*r + 1.0);
    }
    const double q = std::sqrt(-2.0 * std::log(1.0 - p));
    return -(((((c[0]*q + c[1])*q + c[2])*q + c[3])*q + c[4])*q + c[5])
          / ((((d[0]*q + d[1])*q + d[2])*q + d[3])*q + 1.0);
}


// =====================================================================
// Public validators
// =====================================================================

void validate_model_params(double S0, double sigma, double T) {
    if (S0 <= 0.0) {
        throw std::invalid_argument("S0 must be positive");
    }
    if (sigma <= 0.0) {
        throw std::invalid_argument("sigma must be positive");
    }
    if (T <= 0.0) {
        throw std::invalid_argument("T must be positive");
    }
}

void validate_strike(double K) {
    if (K <= 0.0) {
        throw std::invalid_argument("K must be positive");
    }
}

void validate_n_paths(std::size_t n_paths) {
    if (n_paths < 2) {
        throw std::invalid_argument(
            "n_paths must be at least 2 (for sample variance with "
            "Bessel's correction)");
    }
}

void validate_n_steps(std::size_t n_steps) {
    if (n_steps < 1) {
        throw std::invalid_argument("n_steps must be at least 1");
    }
}

void validate_confidence_level(double confidence_level) {
    if (!(confidence_level > 0.0 && confidence_level < 1.0)) {
        throw std::invalid_argument(
            "confidence_level must be in (0, 1)");
    }
}


// =====================================================================
// simulate_terminal_gbm  (Block 1.1)
// =====================================================================

std::vector<double>
simulate_terminal_gbm(double S0, double r, double sigma, double T,
                      std::size_t n_paths,
                      std::mt19937_64& rng) {
    validate_model_params(S0, sigma, T);
    validate_n_paths(n_paths);

    const double drift     = (r - 0.5 * sigma * sigma) * T;
    const double diffusion = sigma * std::sqrt(T);

    std::vector<double> S_T(n_paths);
    for (std::size_t i = 0; i < n_paths; ++i) {
        const double z = standard_normal(rng);
        S_T[i] = S0 * std::exp(drift + diffusion * z);
    }
    return S_T;
}


// =====================================================================
// Euler-Maruyama scheme (Block 1.2.1)
// =====================================================================

std::vector<std::vector<double>>
simulate_path_euler(double S0, double r, double sigma, double T,
                    std::size_t n_steps, std::size_t n_paths,
                    std::mt19937_64& rng) {
    validate_model_params(S0, sigma, T);
    validate_n_steps(n_steps);
    validate_n_paths(n_paths);

    const double h       = T / static_cast<double>(n_steps);
    const double sqrt_h  = std::sqrt(h);
    const double rh      = r * h;

    std::vector<std::vector<double>> paths(
        n_paths, std::vector<double>(n_steps + 1));

    for (std::size_t i = 0; i < n_paths; ++i) {
        paths[i][0] = S0;
        for (std::size_t k = 0; k < n_steps; ++k) {
            const double z  = standard_normal(rng);
            const double dW = sqrt_h * z;
            paths[i][k + 1] = paths[i][k] * (1.0 + rh + sigma * dW);
        }
    }
    return paths;
}


std::vector<double>
simulate_terminal_euler(double S0, double r, double sigma, double T,
                        std::size_t n_steps, std::size_t n_paths,
                        std::mt19937_64& rng) {
    validate_model_params(S0, sigma, T);
    validate_n_steps(n_steps);
    validate_n_paths(n_paths);

    const double h       = T / static_cast<double>(n_steps);
    const double sqrt_h  = std::sqrt(h);
    const double rh      = r * h;

    std::vector<double> S(n_paths, S0);
    for (std::size_t i = 0; i < n_paths; ++i) {
        for (std::size_t k = 0; k < n_steps; ++k) {
            const double z  = standard_normal(rng);
            const double dW = sqrt_h * z;
            S[i] *= (1.0 + rh + sigma * dW);
        }
    }
    return S;
}


// =====================================================================
// Milstein scheme (Block 1.2.2)
// =====================================================================

std::vector<std::vector<double>>
simulate_path_milstein(double S0, double r, double sigma, double T,
                       std::size_t n_steps, std::size_t n_paths,
                       std::mt19937_64& rng) {
    validate_model_params(S0, sigma, T);
    validate_n_steps(n_steps);
    validate_n_paths(n_paths);

    const double h        = T / static_cast<double>(n_steps);
    const double sqrt_h   = std::sqrt(h);
    const double rh       = r * h;
    const double half_s2  = 0.5 * sigma * sigma;

    std::vector<std::vector<double>> paths(
        n_paths, std::vector<double>(n_steps + 1));

    for (std::size_t i = 0; i < n_paths; ++i) {
        paths[i][0] = S0;
        for (std::size_t k = 0; k < n_steps; ++k) {
            const double z  = standard_normal(rng);
            const double dW = sqrt_h * z;
            // factor = 1 + r*h + sigma*dW + 0.5*sigma^2*(dW^2 - h)
            const double factor = 1.0 + rh + sigma * dW
                                + half_s2 * (dW * dW - h);
            paths[i][k + 1] = paths[i][k] * factor;
        }
    }
    return paths;
}


std::vector<double>
simulate_terminal_milstein(double S0, double r, double sigma, double T,
                           std::size_t n_steps, std::size_t n_paths,
                           std::mt19937_64& rng) {
    validate_model_params(S0, sigma, T);
    validate_n_steps(n_steps);
    validate_n_paths(n_paths);

    const double h        = T / static_cast<double>(n_steps);
    const double sqrt_h   = std::sqrt(h);
    const double rh       = r * h;
    const double half_s2  = 0.5 * sigma * sigma;

    std::vector<double> S(n_paths, S0);
    for (std::size_t i = 0; i < n_paths; ++i) {
        for (std::size_t k = 0; k < n_steps; ++k) {
            const double z  = standard_normal(rng);
            const double dW = sqrt_h * z;
            const double factor = 1.0 + rh + sigma * dW
                                + half_s2 * (dW * dW - h);
            S[i] *= factor;
        }
    }
    return S;
}

}  // namespace quant

File: python/quantlib/gbm.py
"""GBM-specific samplers for Monte Carlo simulation.

This module contains all samplers of the geometric Brownian motion
state under the risk-neutral measure, plus the input validators that
are shared across them. It is designed to be paired with
``quantlib.monte_carlo``, which provides the model-agnostic statistical
reducer and the high-level pricers that orchestrate samplers, payoffs
and the reducer.

Public samplers:

- ``simulate_terminal_gbm``: exact sampler of the terminal price
  ``S_T`` under GBM, using the closed-form solution of the SDE
  (Block 1.1).

- ``simulate_path_euler``: Euler-Maruyama path sampler returning the
  full path including the initial value (Block 1.2.1).
- ``simulate_terminal_euler``: convenience that returns only the
  terminal column of ``simulate_path_euler`` (Block 1.2.1).

- ``simulate_path_milstein``: Milstein path sampler returning the
  full path (Block 1.2.2).
- ``simulate_terminal_milstein``: terminal-value convenience for
  Milstein (Block 1.2.2).

Public validators (used by the pricers in ``monte_carlo.py``):

- ``validate_model_params``: checks ``S0 > 0``, ``sigma > 0``, ``T > 0``;
  ``r`` is unconstrained.
- ``validate_strike``: checks ``K > 0``.
- ``validate_n_paths``: checks ``n_paths >= 2``.
- ``validate_n_steps``: checks ``n_steps >= 1``.

References
----------
Phase 2 Block 0 writeup (foundations); Block 1.1 writeup (exact);
Block 1.2.0 writeup (SDE discretisation theory); Block 1.2.1 writeup
(Euler); Block 1.2.2 writeup (Milstein). Glasserman, *Monte Carlo
Methods in Financial Engineering*, Chapters 1, 3, and 6.
"""

import numpy as np
from scipy.stats import norm


# =====================================================================
# Input validators
# =====================================================================

def validate_model_params(S0, r, sigma, T):
    """Validate the model parameters of geometric Brownian motion.

    Note that ``r`` is unconstrained: negative rates are admissible.
    """
    if S0 <= 0:
        raise ValueError(f"S0 must be positive, got {S0}")
    if sigma <= 0:
        raise ValueError(f"sigma must be positive, got {sigma}")
    if T <= 0:
        raise ValueError(f"T must be positive, got {T}")


def validate_strike(K):
    if K <= 0:
        raise ValueError(f"K must be positive, got {K}")


def validate_n_paths(n_paths):
    if not isinstance(n_paths, (int, np.integer)):
        raise TypeError(
            f"n_paths must be an integer, got {type(n_paths).__name__}"
        )
    if n_paths < 2:
        raise ValueError(
            f"n_paths must be at least 2 (for sample variance with "
            f"Bessel's correction), got {n_paths}"
        )


def validate_n_steps(n_steps):
    if not isinstance(n_steps, (int, np.integer)):
        raise TypeError(
            f"n_steps must be an integer, got {type(n_steps).__name__}"
        )
    if n_steps < 1:
        raise ValueError(f"n_steps must be at least 1, got {n_steps}")


# =====================================================================
# Internal helpers
# =====================================================================

def _standard_normals(n, rng):
    """Generate ``n`` independent standard normal samples by inversion.

    Inversion is used (rather than the faster Ziggurat method that
    NumPy provides via ``rng.standard_normal``) to maintain
    coordinate-by-coordinate compatibility with low-discrepancy
    sequences, which will be introduced in Phase 2 Block 3 (QMC).
    See the Phase 2 Block 0 writeup, Section 3.3, for the rationale.
    """
    u = rng.uniform(size=n)
    return norm.ppf(u)


def _gbm_exact_from_brownian(S0, r, sigma, T, W_T):
    """Exact GBM terminal price given the terminal Brownian value(s).

    Computes ``S_T = S_0 * exp((r - 0.5 * sigma^2) * T + sigma * W_T)``,
    the closed-form solution of the GBM SDE evaluated at the supplied
    ``W_T``. Used in validation to build an exact reference path that
    shares its Brownian driver with a discretised path, enabling
    pathwise (strong) and CRN-based (weak) error estimation.

    Notes
    -----
    Private to this module: only the validation scripts import it.
    The function is not part of the user-facing API.
    """
    drift = (r - 0.5 * sigma * sigma) * T
    return S0 * np.exp(drift + sigma * W_T)


# =====================================================================
# Exact GBM sampler (Block 1.1)
# =====================================================================

def simulate_terminal_gbm(S0, r, sigma, T, n_paths, rng):
    """Simulate ``n_paths`` samples of ``S_T`` under geometric Brownian
    motion, exactly.

    Uses the closed-form solution of the GBM SDE,
    ``S_T = S_0 * exp((r - sigma^2/2) * T + sigma * sqrt(T) * Z)``,
    with ``Z ~ N(0, 1)`` sampled by inversion. No time-discretisation
    error.

    See Phase 2 Block 1.1 writeup, Section 3.
    """
    validate_model_params(S0, r, sigma, T)
    validate_n_paths(n_paths)

    Z = _standard_normals(n_paths, rng)
    drift = (r - 0.5 * sigma * sigma) * T
    diffusion = sigma * np.sqrt(T)
    return S0 * np.exp(drift + diffusion * Z)


# =====================================================================
# Internal: sample Brownian increments and resolve rng vs delta_W
# =====================================================================

def _resolve_increments(n_paths, n_steps, h, rng, delta_W):
    """Either generate Brownian increments from rng or validate the
    pre-supplied ones. Used by both Euler and Milstein samplers.

    Exactly one of ``rng`` and ``delta_W`` must be provided.
    """
    if (rng is None) == (delta_W is None):
        raise ValueError(
            "Pass exactly one of `rng` (Generator) or `delta_W` "
            "(pre-sampled increments)."
        )

    if delta_W is None:
        return rng.normal(loc=0.0, scale=np.sqrt(h),
                          size=(n_paths, n_steps))

    delta_W = np.asarray(delta_W, dtype=np.float64)
    if delta_W.shape != (n_paths, n_steps):
        raise ValueError(
            f"delta_W shape must be ({n_paths}, {n_steps}), "
            f"got {delta_W.shape}"
        )
    return delta_W


# =====================================================================
# Euler-Maruyama scheme (Block 1.2.1)
# =====================================================================

def simulate_path_euler(S0, r, sigma, T, n_steps, n_paths,
                        *,
                        rng=None,
                        delta_W=None):
    """Simulate full Euler-Maruyama paths of GBM.

    Implements the recursion
        S_{n+1} = S_n * (1 + r * h + sigma * dW_n),
    with h = T / n_steps and dW_n ~ N(0, h) independent across n.
    Returns the full path including the initial value.

    Parameters
    ----------
    rng : numpy.random.Generator or None, keyword-only
        Random generator for sampling Brownian increments. Pass
        exactly one of ``rng`` and ``delta_W``.
    delta_W : ndarray of shape (n_paths, n_steps) or None, keyword-only
        Pre-sampled Brownian increments. Used by the validation suite
        for common-random-numbers comparisons.

    Returns
    -------
    ndarray of shape (n_paths, n_steps + 1)
        Column 0 is ``S0``, column ``n_steps`` is the
        Euler-discretised terminal value.

    Notes
    -----
    Euler does not preserve positivity in principle: a sufficiently
    negative ``delta_W`` can produce a negative ``S``. For typical
    finance parameters this is astronomically rare and not handled
    here. See Phase 2 Block 1.2.1 writeup, Section 2.2.
    """
    validate_model_params(S0, r, sigma, T)
    validate_n_steps(n_steps)
    validate_n_paths(n_paths)

    h = T / n_steps
    delta_W = _resolve_increments(n_paths, n_steps, h, rng, delta_W)

    # Multiplicative recursion: S_{n+1} = S_n * factor_n
    # where factor_n = 1 + r * h + sigma * delta_W_n.
    factors = 1.0 + r * h + sigma * delta_W
    cumulative = np.cumprod(factors, axis=1)

    paths = np.empty((n_paths, n_steps + 1), dtype=np.float64)
    paths[:, 0] = S0
    paths[:, 1:] = S0 * cumulative
    return paths


def simulate_terminal_euler(S0, r, sigma, T, n_steps, n_paths,
                            *,
                            rng=None,
                            delta_W=None):
    """Convenience: Euler-Maruyama terminal value only.

    Returns only the terminal column of ``simulate_path_euler``. See
    that function for the full parameter description.
    """
    paths = simulate_path_euler(S0, r, sigma, T, n_steps, n_paths,
                                rng=rng, delta_W=delta_W)
    return paths[:, -1]


# =====================================================================
# Milstein scheme (Block 1.2.2)
# =====================================================================

def simulate_path_milstein(S0, r, sigma, T, n_steps, n_paths,
                           *,
                           rng=None,
                           delta_W=None):
    """Simulate full Milstein paths of GBM.

    Implements the recursion
        S_{n+1} = S_n * (1 + r*h + sigma*dW_n
                          + 0.5 * sigma^2 * (dW_n^2 - h)),
    with h = T / n_steps and dW_n ~ N(0, h) independent across n.

    The recursion is multiplicative but quadratic in dW_n. The extra
    term, called the Milstein corrector, has zero mean (because
    E[dW_n^2 - h] = 0 since Var(dW_n) = h) but is pathwise nonzero
    of typical magnitude h. Including it lifts the strong order from
    1/2 (Euler) to 1 (Milstein); the weak order remains 1.

    Parameters and return value identical to ``simulate_path_euler``.

    Notes
    -----
    Milstein has marginally better positivity behaviour than Euler:
    the squared-increment term is always nonnegative, so the
    multiplicative factor is bounded below by
    ``1 + r*h + sigma*dW_n - 0.5 * sigma^2 * h``. For typical finance
    parameters, neither scheme produces nonpositive prices.

    See Phase 2 Block 1.2.2 writeup.
    """
    validate_model_params(S0, r, sigma, T)
    validate_n_steps(n_steps)
    validate_n_paths(n_paths)

    h = T / n_steps
    delta_W = _resolve_increments(n_paths, n_steps, h, rng, delta_W)

    # Multiplicative recursion with quadratic corrector.
    # factor_n = 1 + r*h + sigma * dW_n + 0.5 * sigma^2 * (dW_n^2 - h)
    sigma_sq = sigma * sigma
    factors = (1.0 + r * h
               + sigma * delta_W
               + 0.5 * sigma_sq * (delta_W * delta_W - h))
    cumulative = np.cumprod(factors, axis=1)

    paths = np.empty((n_paths, n_steps + 1), dtype=np.float64)
    paths[:, 0] = S0
    paths[:, 1:] = S0 * cumulative
    return paths


def simulate_terminal_milstein(S0, r, sigma, T, n_steps, n_paths,
                               *,
                               rng=None,
                               delta_W=None):
    """Convenience: Milstein terminal value only.

    Returns only the terminal column of ``simulate_path_milstein``.
    See that function for the full parameter description.
    """
    paths = simulate_path_milstein(S0, r, sigma, T, n_steps, n_paths,
                                   rng=rng, delta_W=delta_W)
    return paths[:, -1]
